File: Java/Tema5_Ejercicio5.java
/*
 * TEMA 5: ARREGLOS Y LISTAS - EJERCICIO 5
 * ========================================
 * Escriba un programa que invierta un arreglo.
 */

import java.util.Arrays;

public class Tema5_Ejercicio5 {
    public static void main(String[] args) {
        int[] numeros = {10, 20, 30, 40, 50};
        
        System.out.println("=== ARREGLO ORIGINAL ===");
        System.out.println(Arrays.toString(numeros));
        
        // Método 1: Invertir con bucle
        int[] invertido = new int[numeros.length];
        for (int i = 0; i < numeros.length; i++) {
            invertido[i] = numeros[numeros.length - 1 - i];
        }
        
        System.out.println("\n=== ARREGLO INVERTIDO (copia) ===");
        System.out.println(Arrays.toString(invertido));
        
        // Método 2: Invertir in-situ
        int[] numeros2 = {10, 20, 30, 40, 50};
        for (int i = 0; i < numeros2.length / 2; i++) {
            int temp = numeros2[i];
            numeros2[i] = numeros2[numeros2.length - 1 - i];
            numeros2[numeros2.length - 1 - i] = temp;
        }
        
        System.out.println("\n=== ARREGLO INVERTIDO (in-situ) ===");
        System.out.println(Arrays.toString(numeros2));
    }
}

/*
 * SALIDA ESPERADA:
 * === ARREGLO ORIGINAL ===
 * [10, 20, 30, 40, 50]
 * 
 * === ARREGLO INVERTIDO (copia) ===
 * [50, 40, 30, 20, 10]
 * 
 * === ARREGLO INVERTIDO (in-situ) ===
 * [50, 40, 30, 20, 10]
 */

File: JS/Tema5_Ejercicio5.js
/*
 * TEMA 5: ARREGLOS Y LISTAS - EJERCICIO 5
 * ========================================
 * Invertir un array (in-place y copia).
 * 
 * EXPLICACIÓN:
 * - Método reverse() modifica el original
 * - Crear copia invertida
 * - Dos enfoques diferentes
 */

function invertirInPlace(arr) {
    return arr.reverse();
}

function invertirCopia(arr) {
    return [...arr].reverse();
}

function invertirManual(arr) {
    const resultado = [];
    for (let i = arr.length - 1; i >= 0; i--) {
        resultado.push(arr[i]);
    }
    return resultado;
}

const original = [1, 2, 3, 4, 5];

console.log("=== INVERTIR ARRAYS ===");
console.log(`Array original: ${original}`);

const copia1 = [...original];
const invertida1 = invertirInPlace(copia1);
console.log(`Con reverse(): ${invertida1}`);
console.log(`Original modificado: ${copia1}`);

const invertida2 = invertirCopia(original);
console.log(`\nCopia invertida: ${invertida2}`);
console.log(`Original intacto: ${original}`);

const invertida3 = invertirManual(original);
console.log(`\nManualmente: ${invertida3}`);

/*
 * SALIDA ESPERADA:
 * === INVERTIR ARRAYS ===
 * Array original: 1,2,3,4,5
 * Con reverse(): 5,4,3,2,1
 * Original modificado: 5,4,3,2,1
 * 
 * Copia invertida: 5,4,3,2,1
 * Original intacto: 1,2,3,4,5
 * 
 * Manualmente: 5,4,3,2,1
 */
